internal-dns/src/resolver.rs

```diff
@@ -287,6 +276,55 @@ impl Resolver {
         })
     }
 
+    /// Returns the targets of the SRV records for a DNS name.
+    ///
+    /// Unlike [`Resolver::lookup_srv`], this will further lookup the returned
+    /// targets and return a list of [`SocketAddrV6`].
+    pub async fn lookup_all_socket_v6(
+        &self,
+        service: crate::ServiceName,
+    ) -> Result<Vec<SocketAddrV6>, ResolveError> {
+        let name = service.srv_name();
+        trace!(self.log, "lookup_all_socket_v6 srv"; "dns_name" => &name);
+        let response = self.resolver.srv_lookup(&name).await?;
+        debug!(
+            self.log,
+            "lookup_all_socket_v6 srv";
+            "dns_name" => &name,
+            "response" => ?response
+        );
+
+        // See `lookup_all_ipv6` for more details, but we need to do
+        // an extra lookup per-target here so we can construct a bunch of
+        // futures to query them concurrently.
+        let log = &self.log;
+        let futures = response.iter().map(|srv| async {
+            let target = srv.target();
+            let port = srv.port();
+            trace!(
+                log,
+                "lookup_all_socket_v6: looking up SRV target";
+                "name" => ?target,
+            );
+            self.resolver
+                .ipv6_lookup(target.clone())
+                .await
+                .map(|ips| (ips, port))
+        });
+        let results = futures::future::try_join_all(futures).await?;
```

Comment:
Are we sure `self.resolver.ipv6_lookup` is cancel-safe?

Related: Do we actually *want* to exit early here? Or would we want to return all the futures that can return something successful? 

(Just comparing the usage here with https://docs.rs/cancel-safe-futures/0.1.2/cancel_safe_futures/macro.join_then_try.html )

Reply:
Good catch! This was basically a counterpart to the existing `lookup_all_ipv6` method and based on the comment we definitely want all the ones that are successful:
```
        // What do we do if some of these queries succeed while others fail?  We
        // may have some addresses, but the list might be incomplete.  That
        // might be okay for some use cases but not others.  For now, we do the
        // simple thing.  In the future, we'll want a more cueball-like resolver
        // interface that better deals with these cases.
```

Reply:
Moved that logic out to a separate helper and fixed in #3785.